Use upstream chart for external-dns for AWS integration

abaguas opened the pull request:
The latest version of the external-dns image requires setting the AWS_DEFAULT_REGION environment variable.
The introduction of this variable demands changes in our chart as well as changes to the user configuration. Therefore, it is a good opportunity to do breaking changes and introduce the upstream external-dns helm chart.
By using the upstream chart we can profit from all the features of the external dns community out-of-the-box. We no longer have to develop a wrapper on our side, which is currently the bottleneck to supporting further upstream DNS providers. These features include both authentication options and new upstream providers.

The change deprecates only the route53 integration, all other integrations are not affected.

## Review tips

Checkout the first commit of this PR, fb11af4, and compare the route53 integration using the old route53 integration, or the new extdns integration:
`helm package -u . > /dev/null && helm template k8gb "k8gb-v0.14.0.tgz" --include-crds -f values.yaml -f values-route53.yaml > route53.yaml`
`helm package -u . > /dev/null && helm template k8gb "k8gb-v0.14.0.tgz" --include-crds -f values.yaml -f values-extdns.yaml > extdns.yaml`

The second commit removes the route53 configuration from the chart.

## Required user changes

The following configuration changes should be included in the release notes:

#### Auth configuration
1.
```
route53:
  enabled: true
  secret: credentials
```
becomes
```
extdns:
  enabled: true
  provider:
    name: aws
  extraVolumes:
  - name: aws-credentials
    secret:
      secretName: credentials
  extraVolumeMounts:
  - name: aws-credentials
    mountPath: /.aws
    readOnly: true
```

2.
```
route53:
  enabled: true
  irsaRole: arn:aws:iam::111111:role/external-dns
```
becomes:
```
extdns:
  enabled: true
  provider:
    name: aws
  serviceAccount:
    annotations:
      eks.amazonaws.com/role-arn: arn:aws:iam::111111:role/external-dns
```

3.
```
route53:
  enabled: true
  assumeRoleArn: role
```
becomes:
```
extdns:
  enabled: true
  provider:
    name: aws
  extraArgs:
    aws-assume-role: role
```

#### General configuration
In addition, the AWS_DEFAULT_REGION must be specified using an environment variable, for example:
```
extdns:
  env:
  - name: AWS_DEFAULT_REGION
    value: "us-east-1"
```
And a couple of variables must be manually specified (there are helm validation function that make sure they are correct). Replace `<GEOTAG>` with the same value as `k8gb.clusterGeoTag`, and `domainFilters` with the same values as `k8gb.dnsZones.zone`/`k8gb.dnsZone`:
```
extdns:
  txtPrefix: "k8gb-<GEOTAG>"
  txtOwnerId: "k8gb-<GEOTAG>"
  domainFilters:
  - "<domain>"
```
Note: if you used to set `hostedZoneID`, then txtOwnerId will take the value `k8gb-<hostZoneID>-<GEOTAG>`.

k0da commented on chart/k8gb/templates/external-dns/dns-endpoint-crd-manifest.yaml:
```diff
@@ -1,3 +1,4 @@
+{{- if not .Values.extdns.enabled }}
```
CRD presence is not dependent on whether external-dns is used or not. DNSEndpoint is always used as source record for CoreDNS application records

abaguas replied:
The upstream external dns chart installs the same DNSEndpoint CRD, the goal of this condition is not to generate the CRD twice.
https://github.com/kubernetes-sigs/external-dns/blob/master/charts/external-dns/crds/dnsendpoint.yaml

ytsarev commented on chart/k8gb/templates/deployment.yaml:
```diff
@@ -104,7 +107,7 @@ spec:
                   key: INFOBLOX_WAPI_PASSWORD
             {{- end }}
-            {{- if or .Values.route53.enabled .Values.ns1.enabled .Values.rfc2136.enabled .Values.azuredns.enabled .Values.cloudflare.enabled }}
+            {{- if or .Values.extdns.enabled .Values.ns1.enabled .Values.rfc2136.enabled .Values.azuredns.enabled .Values.cloudflare.enabled }}
```
My first impression is that it does not look consistent. Why are we using the chart only for route53 and not for everything else?

ytsarev replied:
@abaguas it still looks quite complex. I will create alternative PR with straightforward templating, then we can compare and decide how to proceed :)

abaguas replied:
What looks complex? This file in particular?

abaguas replied:
For illustration purposes I added commit 1607524 which removes all the configuration no longer necessary by enabling the external-dns helm chart

ytsarev replied:
@abaguas convinced :) Maintaining the custom templates is more of an effort in any case. Let's proceed with the switch. Do you want to proceed gradually or switch them all in 1 PR?

abaguas replied:
I prefer to switch gradually, 1 provider per PR. It will make it easier to verify functionality and help users with migration instructions. I will remote the last commit

ytsarev reviewed (commented):
Tried to test the change e2e

* Deployed local setup
```
K8GB_LOCAL_VERSION=test make deploy-full-local-setup
```

* Tried example extdns config
```
 helm -n k8gb upgrade -i -f values-extdns.yaml k8gb .
```
*  Hit CrashLoopBackOff with failing depresolver 

```
 k get pod -A
NAMESPACE         NAME                                           READY   STATUS              RESTARTS        AGE
istio-ingress     istio-ingressgateway-694bcd8d4c-nk78b          1/1     Running             0               115m
istio-system      istiod-6c854cfc5d-bfcrd                        1/1     Running             0               115m
k8gb              k8gb-7b8767696c-5msmx                          0/1     CrashLoopBackOff    5 (2m19s ago)   5m18s
```

```
 k -n k8gb logs -f k8gb-7b8767696c-5msmx
2025-03-26T22:32:17Z INF github.com/k8gb-io/k8gb/controllers/logging/log.go:71 > Logger configured
2025-03-26T22:32:17Z INF github.com/k8gb-io/k8gb/controllers/logging/log.go:75 > Logger settings format=simple
2025-03-26T22:32:17Z INF ./main.go:75 > k8gb info commit=1874f6c95a19f290455bf5a85ef7c3ac04c8d4e2 version=v0.14.0
2025-03-26T22:32:17Z ERR ./main.go:77 > Can't resolve environment variables error="'EDGE_DNS_ZONE' is empty"
```

@kuritka It is probably related to the recent enhancement and refactoring to support multiple domains

kuritka commented:
@ytsarev , I see, wil llook on that.

abaguas commented:
> Tried to test the change e2e
> 
> * Deployed local setup
> 
> ```
> K8GB_LOCAL_VERSION=test make deploy-full-local-setup
> ```
> 
> * Tried example extdns config
> 
> ```
>  helm -n k8gb upgrade -i -f values-extdns.yaml k8gb .
> ```
> 
> * Hit CrashLoopBackOff with failing depresolver
> 
> ```
>  k get pod -A
> NAMESPACE         NAME                                           READY   STATUS              RESTARTS        AGE
> istio-ingress     istio-ingressgateway-694bcd8d4c-nk78b          1/1     Running             0               115m
> istio-system      istiod-6c854cfc5d-bfcrd                        1/1     Running             0               115m
> k8gb              k8gb-7b8767696c-5msmx                          0/1     CrashLoopBackOff    5 (2m19s ago)   5m18s
> ```
> 
> ```
>  k -n k8gb logs -f k8gb-7b8767696c-5msmx
> 2025-03-26T22:32:17Z INF github.com/k8gb-io/k8gb/controllers/logging/log.go:71 > Logger configured
> 2025-03-26T22:32:17Z INF github.com/k8gb-io/k8gb/controllers/logging/log.go:75 > Logger settings format=simple
> 2025-03-26T22:32:17Z INF ./main.go:75 > k8gb info commit=1874f6c95a19f290455bf5a85ef7c3ac04c8d4e2 version=v0.14.0
> 2025-03-26T22:32:17Z ERR ./main.go:77 > Can't resolve environment variables error="'EDGE_DNS_ZONE' is empty"
> ```
> 
> @kuritka It is probably related to the recent enhancement and refactoring to support multiple domains

I rebased on Michal's fix and changed the zone to `k8gb.io` in the `values-extdns.yaml` since AWS doesn't let us create the `example.com`.
Testing the change is not as easy as running a helm upgrade --install on the local setup because you will get the last released image. The example commands I included in the commit message only show the generated manifests with the old route53 configuration and the new one are equivalent. But that is not enough to run in AWS.

Here is how you can test:
```
# create the cluster
export CHART=chart/k8gb
export VALUES_YAML=chart/k8gb/values-extdns.yaml
K8GB_LOCAL_VERSION=test FULL_LOCAL_SETUP_WITH_APPS=false make deploy-full-local-setup

# create the AWS secret manually and apply them to the cluster
cat <<-EOF > credentials
[default]
aws_access_key_id = $(echo $SECRET_ACCESS_KEY | jq -r '.AccessKey.AccessKeyId')
aws_secret_access_key = $(echo $SECRET_ACCESS_KEY | jq -r '.AccessKey.SecretAccessKey')
EOF

kubectl create secret generic credentials -n k8gb --from-file credentials --context k3d-test-gslb1
kubectl create secret generic credentials -n k8gb --from-file credentials --context k3d-test-gslb2
```

You will see the NS records being created. Then, once you apply a k8gb resource you will also see its A records.

Note: the old extdns for rfc will still be deployed on the cluster `external-dns-5dd8485c8-s48cm` and will crash loop. But you can ignore it.

abaguas commented on Makefile:
```diff
@@ -547,7 +547,7 @@ $(shell echo $(foreach cl,$(filter-out $1,$(shell seq $(CLUSTERS_NUMBER))),$(cal
 
 define get-helm-args
-k8gb.clusterGeoTag='$(call nth-geo-tag,$1)' --set k8gb.extGslbClustersGeoTags='$(call get-ext-tags,$1)'
+k8gb.clusterGeoTag='$(call nth-geo-tag,$1)' --set k8gb.extGslbClustersGeoTags='$(call get-ext-tags,$1)' --set extdns.txtPrefix='k8gb-$(call nth-geo-tag,$1)-' --set extdns.txtOwnerId='k8gb-$(call nth-geo-tag,$1)'
```
@ytsarev The `extdns` I added here are breaking upgrade testing (not compliant with the old `values.schema.json`). I will remove them once you test e2e

ytsarev reviewed (approved):
@abaguas i rebased branch with the latest `master` and tested it e2e - everything works as expected, thank you so much!

I pushed rebased branch to the PR branch so we have up2date tested state here

abaguas commented:
Thank you! I am going to work now on the automated script to test this
